modules/data_dictionary_widget/src/Plugin/Field/FieldWidget/DataDictionaryWidget.php

```diff
@@ -0,0 +1,577 @@
+   */
+  public function formElement(FieldItemListInterface $items, $delta, array $element, array &$form, FormStateInterface $form_state) {
+    $field_values = $form_state->getValue(["field_json_metadata"]);
+    $current_fields = $form_state->get('current_fields');
+    $fields_being_modified = $form_state->get("fields_being_modified") ?? NULL;
+    $op = $form_state->getTriggeringElement()['#op'] ?? NULL;
+    $field_json_metadata = !empty($items[0]->value) ? json_decode($items[0]->value, TRUE) : [];
+    $op_index = isset($form_state->getTriggeringElement()['#op']) ? explode("_", $form_state->getTriggeringElement()['#op']) : NULL;
+
+    $data_results = $field_json_metadata ? $field_json_metadata["data"]["fields"] : [];
+
+    // Build the data_results array to display the rows in the data table.
+    $data_results = $this->processDataResults($data_results, $current_fields, $field_values, $op);
+
+    $element['identifier'] = [
+      '#type' => 'textfield',
+      '#required' => TRUE,
+      '#title' => $this->t('Identifier'),
+      '#default_value' => $field_json_metadata['identifier'] ?? '',
+    ];
+
+    $element['title'] = [
+      '#type' => 'textfield',
+      '#required' => TRUE,
+      '#title' => $this->t('Title'),
+      '#default_value' => $field_json_metadata['title'] ?? '',
+    ];
+
+    $element['dictionary_fields'] = [
+      '#type' => 'fieldset',
+      '#title' => $this->t('Data Dictionary Fields'),
+      '#prefix' => '<div id = field-json-metadata-dictionary-fields>',
+      '#suffix' => '</div>',
+      '#markup' => $this->t('<div class="claro-details__description">A data dictionary for this resource, compliant with the <a href="https://specs.frictionlessdata.io/table-schema/" target="_blank">Table Schema</a> specification.</div>'),
+      '#pre_render' => [
+      [$this, 'preRenderForm'],
+      ],
+    ];
+
+    $element['dictionary_fields']['data'] = [
+      '#access' => ((bool) $current_fields || (bool) $data_results),
+      '#type' => 'table',
+      '#header' => ['NAME', 'TITLE', 'DETAILS'],
+      '#rows' => $form_state->get('cancel') ? $current_fields : ($data_results ?? []),
+      '#tree' => TRUE,
+      '#theme' => 'custom_table',
+    ];
+    $action_list = [
+      'format',
+      'edit',
+      'update',
+      'abort',
+      'delete',
+    ];
+
+    // Creating ajax buttons/fields to be placed in correct location later.
+    foreach ($data_results as $key => $data) {
+      if (isset($op_index[0]) && in_array($op_index[0], $action_list) && array_key_exists($key, $fields_being_modified)) {
+        $element['dictionary_fields']['edit_fields'][$key]['name'] = [
+          '#name' => 'field_json_metadata[0][dictionary_fields][data][' . $key . '][field_collection][name]',
+          '#type' => 'textfield',
+          '#value' => $this->t($current_fields[$key]['name']),
+          '#required' => TRUE,
+          '#title' => 'Name',
+          '#description' => $this->t('A name for this field.'),
```

Comment:
'#description' => $this->t('Machine name of the field/column in the data table.'),